app/assets/javascripts/admin/addon/components/admin-report.gjs

```diff
@@ -0,0 +1,793 @@
+
+    let reportKey = "reports:";
+    reportKey += [
+      this.args.dataSourceName,
+      isTesting() ? "start" : formattedStartDate.replace(/-/g, ""),
+      isTesting() ? "end" : formattedEndDate.replace(/-/g, ""),
+      "[:prev_period]",
+      this.args.reportOptions?.table?.limit,
+      // Convert all filter values to strings to ensure unique serialization
+      this.args.filters?.customFilters
+        ? JSON.stringify(this.args.filters?.customFilters, (k, v) =>
+            k ? `${v}` : v
+          )
+        : null,
+      SCHEMA_VERSION,
+    ]
+      .filter((x) => x)
+      .map((x) => x.toString())
+      .join(":");
+
+    return reportKey;
+  }
+
+  get chartGroupings() {
+    const chartGrouping = this.options?.chartGrouping;
+    const options = ["daily", "weekly", "monthly"];
+
+    return options.map((id) => {
+      return {
+        id,
+        disabled:
+          id === "daily" && this.model.chartData.length >= DAILY_LIMIT_DAYS,
+        label: `admin.dashboard.reports.${id}`,
+        class: `chart-grouping ${chartGrouping === id ? "active" : "inactive"}`,
+      };
+    });
+  }
+
+  @action
+  onChangeDateRange(range) {
+    this.dateRangeFrom = range.from;
+    this.dateRangeTo = range.to;
+  }
+
+  @action
+  applyFilter(id, value) {
+    let customFilters = this.args.filters?.customFilters || {};
+
+    if (typeof value === "undefined") {
+      delete customFilters[id];
+    } else {
+      customFilters[id] = value;
+    }
+
+    this.refreshReport({ filters: customFilters });
+  }
+
+  @action
+  refreshReport(options = {}) {
+    if (!this.args.onRefresh) {
+      return;
+    }
+
+    this.args.onRefresh({
+      type: this.model.type,
+      mode: this.currentMode,
+      chartGrouping: options.chartGrouping,
+      startDate:
+        typeof options.startDate === "undefined"
+          ? this.startDate
+          : options.startDate,
+      endDate:
+        typeof options.endDate === "undefined" ? this.endDate : options.endDate,
+      filters:
+        typeof options.filters === "undefined"
+          ? this.args.filters?.customFilters
+          : options.filters,
+    });
+  }
+
+  @action
+  exportCsv() {
+    const args = {
+      name: this.model.type,
+      start_date: this.startDate.toISOString(true).split("T")[0],
+      end_date: this.endDate.toISOString(true).split("T")[0],
+    };
+
+    const customFilters = this.args.filters?.customFilters;
+    if (customFilters) {
+      Object.assign(args, customFilters);
+    }
+
+    exportEntity("report", args).then(outputExportResult);
+  }
+
+  @action
+  onChangeMode(mode) {
+    this.currentMode = mode;
+    this.refreshReport({ chartGrouping: null });
+  }
+
+  @bind
+  fetchOrRender() {
+    if (this.report) {
+      this._renderReport(this.report, this.forcedModes, this.currentMode);
+    } else if (this.args.dataSourceName) {
+      this._fetchReport();
+    }
+  }
+
+  @bind
```

Comment:
no need for bind?

Reply:
I had to use bind on this and the other 2 places because when I called the main function from `didUpdate` `this` did not exist on any of these funcitons